db/compaction/compaction_iterator.cc

```diff
@@ -1075,6 +1075,52 @@ void CompactionIterator::GarbageCollectBlobIfNeeded() {
   }
 }
 
+void CompactionIterator::DecideOutputLevel() {
+#ifndef NDEBUG
+  // TODO: will be set by sequence number or key range, for now, it will only be
+  // set by unittest
+  PerKeyPlacementContext context(level_, ikey_.user_key, value_,
+                                 ikey_.sequence);
+  TEST_SYNC_POINT_CALLBACK("CompactionIterator::PrepareOutput.context",
+                           &context);
+  output_to_penultimate_level_ = context.output_to_penultimate_level;
+#endif /* !NDEBUG */
+
+  // if the key is within the earliest snapshot, it has to output to the
+  // penultimate level.
+  if (ikey_.sequence > earliest_snapshot_) {
+    output_to_penultimate_level_ = true;
+  }
+
+  if (output_to_penultimate_level_) {
+    // If it's decided to output to the penultimate level, but unsafe to do so,
+    // still output to the last level. For example, moving the data from a lower
+    // level to a higher level outside of the higher-level input key range is
+    // considered unsafe, because the key may conflict with higher-level SSTs
+    // not from this compaction.
+    // TODO: add statistic for declined output_to_penultimate_level
+    bool safe_to_penultimate_level =
+        compaction_->WithinPenultimateLevelOutputRange(ikey_.user_key);
+    if (!safe_to_penultimate_level) {
+      output_to_penultimate_level_ = false;
+      // It could happen when disable/enable `bottommost_temperature` while
+      // holding a snapshot. When `bottommost_temperature` is not set
+      // (==kUnknown), the data newer than any snapshot is pushed to the last
+      // level, but when the per_key_placement feature is enabled on the fly,
+      // the data later than the snapshot has to be moved to the penultimate
```

Comment:
Does the data later than a snapshot have to be moved to the penultimate level? Shouldn't it be OK for the penultimate level to reject it and leave it in the last level?

Reply:
My memory was vague. Was it because of a discussion related to range delete?

Reply:
Not sure. A wide range tombstone in the penultimate level (i.e., one that spans beyond the penultimate level input files' bounds) forces compaction to the last level - https://github.com/facebook/rocksdb/blob/32520df1d9ab02efe9465af50ce077d8539946c4/db/compaction/compaction.cc#L383

Actually, that is what caused me to run into this `Status::Corruption` in the first place: a penultimate level had a wide range tombstone so was forced to write its output all to the last level, while some of it was newer than a snapshot. However I think it looked safe (in terms of correctness) to write it all in the last level.

Reply:
> a penultimate level had a wide range tombstone so was forced to write its output all to the last level, while some of it was newer than a snapshot. However I think it looked safe (in terms of correctness) to write it all in the last level

In that case it is. But Siying explained a case where it would not be:

- A range deletion is preserved in penultimate level because it's newer than the earliest snapshot
- A key newer than the range deletion (and thus also newer than the earliest snapshot) is output to the last level. This is what the `Status::Corruption` here prevents.
- Then the range deletion would accidentally cover a newer key

Reply:
How does point 1 and point 2 happen? I assume point 1 can happen when a range tombstone is compacted while per_key_placement is enabled. Point 2 happens when `per_key_placement` is disabled, but when it is compacted down to the last level, the range tombstone should be compacted together with it. Is the range deletion from point 1 moved from last level to the penultimate level?

Reply:
@cbi42 Yes point 1 happens whenever a range tombstone is involved in a last-level compaction and is not obsoleted. If not for the `Status::Corruption()`, point 2 would happen when `WithinPenultimateLevelOutputRange()` returns false for a key newer than a snapshot. That seems impossible unless the preclude last level feature is dynamically changed.